compiler/damlc/lib/DA/Cli/Damlc/DependencyDb.hs

```diff
@@ -246,24 +279,143 @@ validatePkgId pkgId = do
+    , pkgId :: String
+    } deriving Generic
+instance Aeson.FromJSON DependencyInfo
+instance Aeson.ToJSON DependencyInfo
+
+resolvePkgs :: NormalizedFilePath -> Options -> [FullPkgName] -> IO (M.Map FullPkgName LF.PackageId)
+resolvePkgs projRoot opts pkgs
+    | null pkgs = pure M.empty
+    | otherwise = do
+        mbRes <- resolvePkgsWithLockFile lockFp pkgs
+        resOrErr <- case mbRes of
+          Nothing -> resolvePkgsWithLedger depsDir pkgs
+          Just res -> pure $ Right res
+        case resOrErr of
+            Left missing ->
+                fail $
+                unlines $
+                "Unable to resolve the following packages in daml.yaml:" :
+                [ (T.unpack $ LF.unPackageName pkgName) <> "-" <>
+                (T.unpack $ LF.unPackageVersion pkgVersion)
+                | FullPkgName {pkgName, pkgVersion} <- missing
+                ]
+            Right result -> do
+              writeLockFile lockFile result
+              pure result
+  where
+    depsDir = dependenciesDir opts projRoot
+    lockFp = fromNormalizedFilePath projRoot </> lockFile
+
+writeLockFile :: FilePath -> M.Map FullPkgName LF.PackageId -> IO ()
+writeLockFile lockFp resolvedPkgs = do
+    Yaml.encodeFile lockFp $
+        LockFile
+            [ DependencyInfo
+                { name = T.unpack $ LF.unPackageName pkgName
+                , version = T.unpack $ LF.unPackageVersion pkgVersion
+                , pkgId = T.unpack $ LF.unPackageId pkgId
+                }
+            | (FullPkgName {pkgName, pkgVersion}, pkgId) <- M.toList resolvedPkgs
+            ]
+
+-- | Read the package lock file and resolve the given packages. Returns a list of unresolved
+-- packages together with the resolved ones.
+resolvePkgsWithLockFile :: FilePath -> [FullPkgName] -> IO (Maybe (M.Map FullPkgName LF.PackageId))
+resolvePkgsWithLockFile lockFp pkgs = do
+    hasLockFile <- doesFileExist lockFp
+    if hasLockFile
+        then do
+            errOrLock <- Yaml.decodeFileEither lockFp
+            case errOrLock of
+                Left err ->
+                    fail $
+                    "Failed to parse daml.lock: " <> show err <> "\nTry to delete it an run again."
+                Right LockFile {dependencies} -> do
+                    let m =
+                            M.fromList
+                                [ ( FullPkgName
+                                        { pkgName = LF.PackageName $ T.pack $ name d
+                                        , pkgVersion = LF.PackageVersion $ T.pack $ version d
+                                        }
+                                  , LF.PackageId $ T.pack $ pkgId d)
+                                | d <- dependencies
+                                ]
+                    pure $
+                        if Set.fromList pkgs `Set.isSubsetOf` M.keysSet m
+                            then Just $ M.restrictKeys m (Set.fromList pkgs)
+                            else Nothing
```

Comment:
What is this checking for?

Reply:
this checks, that all required packages could be resolved with the lock-file.

Reply:
Maybe worth adding an inline comment here.

Reply:
added an inline comment.